File: 0525-contiguous-array/0525-contiguous-array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        n = len(nums)
        maxLength =0
        sum =0
        mp={}
        mp[0]=-1

        for i in range(n):
            if nums[i]==0:
                sum +=-1
            else:
                sum+=nums[i]
            
            if sum in mp:
                maxLength = max(maxLength,i-mp[sum])
            if sum not in mp:
                mp[sum]=i
        return maxLength

File: 0525-contiguous-array/0525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int n = nums.size();
        int maxLength = 0;
        int sum =0;

        unordered_map<int,int> mp; // <sum-> index>
        mp[0]=-1;

        for(int i =0;i<n;i++){
            if(nums[i]==0){
                sum+=-1;
            }else{
                sum+=nums[i];
            }

            if(mp.find(sum)!=mp.end()){
                maxLength = max(maxLength,i- mp[sum]);
            }
            if(mp.find(sum)==mp.end()){
                mp[sum]=i;
            }
        }
        return maxLength;
    }
};